plugins/api/api_main.ml

```diff
@@ -1,39 +1,215 @@
-  match String.split ~on:'/' name with
-  | ["api"; lang; _] -> lang = language
-  | _ -> false
+module Path = FilePath.DefaultPath.Abstract
+type path = Path.filename
 
-let (/) = Filename.concat
+type api_descr = {
+  lang : string;
+  name : string;
+}
 
-let get_file bundle lang filename =
-  let name = Filename.temp_file "api" "lang" in
-  let file = "api" / lang / filename in
-  info "applying %s" file;
-  match Bundle.get_file ~name bundle (Uri.of_string file) with
-  | None -> None
-  | Some uri -> Some (Uri.path uri)
+type api = {
+  desc : api_descr;
+  path : path;
+}
 
-let files bundle language  =
-  Bundle.list bundle |> List.filter ~f:(in_language ~language) |>
-  List.map ~f:Filename.basename
+module Api_path = struct
 
-let mapper bundle (module Api : Api) =
-  let get_file = get_file bundle Api.language in
-  let files = files bundle Api.language in
+  type t = path
+
+  let (/) = Path.concat
+  let of_string x = Path.reduce @@ Path.filename_of_string x
+  let to_string = Path.string_of_filename
+  let basename = Path.basename
+  let basename' x = to_string @@ basename @@ of_string x
+
+  let exists p = FileUtil.(test Exists) @@ to_string p
+  let is_apiable = FileUtil.(Or (Is_file, Is_link))
+
+  let is_apiable_path p =
+    exists p && FileUtil.test is_apiable @@ to_string p
+
+  let find_files path =
+    try_with_default ~default:[]
+      (fun () -> FileUtil.find is_apiable (to_string path)
+          (fun acc f -> of_string f :: acc) [])
+
+  let path_of_desc desc = of_string desc.lang / of_string desc.name
+  let with_lang path lang = path / of_string lang
+
+  let lang_of_path path =
+    to_string Path.(path |> reduce |> dirname |> basename)
+
+  let name_of_path p = to_string @@ Path.basename p
+  let desc_of_path p = {name = name_of_path p; lang = lang_of_path p;}
+
+  let api_of_path path =
+    if is_apiable_path path then
+      Some {desc = desc_of_path path; path;}
+    else None
+
+  let is_api_at_path desc path = is_apiable_path @@ path / path_of_desc desc
+  let find_all_paths paths desc = List.filter ~f:(is_api_at_path desc) paths
+  let find_path paths desc = List.find ~f:(is_api_at_path desc) paths
+
+  let find_apis ?lang path =
+    let path' = match lang with
+      | Some lang -> with_lang path lang
+      | None -> path in
+    find_files path' |>
+    List.filter_map ~f:api_of_path
+
+end
+
+module Api = struct
+
+  type t = api
+  type op_err = [ `No_dst | `Fail of exn ]
+
+  let lang t = t.desc.lang
+  let name t = t.desc.name
+
+  let create lang path =
+    let path = Api_path.of_string path in
+    if Api_path.is_apiable_path path then
+      let name = Api_path.name_of_path path in
+      Some {desc = {lang; name}; path;}
+    else None
+
+  let cp dest api =
+    let dest' = Api_path.with_lang dest (lang api) in
+    ok_if_true (Api_path.exists dest) ~error:`No_dst >>= fun () ->
+    try_with
+      (fun () ->
+         let dest' = Api_path.to_string dest' in
+         FileUtil.mkdir dest';
```

Comment:
Also, if you're still going to create a path, then you should use `~parent:true` option, to ensure, that you are able to create the path, if the parents do not exist.